Benchmarking sessions in Django

neopostmodern opened the pull request.

Commit:
feat: implement basic CORS headers

- warning: set to allow all hosts currently

Commit:
feat: basic benchmarking

- create benchmarking session with parameters
- run benchmarks in background
- observe running benchmarks
- fetch results with minimal metrics

Commit:
format: lint auto-created migration

Commit:
Merge branch 'refactor-to-django' into django-benchmarking-session

# Conflicts:
#	mmvb_backend/settings.py
#	requirements.txt

Commit:
fix: explicit intermediate status for running benchmarking sessions

Commit:
fix: basic validation of AI implementation response (structure)

Commit:
style: ran black

Commit:
Upgrades pre-commit

maccinza commented on mmvb_backend/benchmarking_sessions/api/views.py:
```diff
@@ -0,0 +1,124 @@
+
+# todo: properly document endpoints
+class BenchmarkingSessionViewSet(ModelViewSet):
+    schema = AutoSchema(tags=["BenchmarkingSessions",])
```
I believe we should use `CamelCaseAutoSchema` from `common.utils`.

maccinza commented on mmvb_backend/benchmarking_sessions/api/views.py:
```diff
@@ -0,0 +1,124 @@
+    # todo: generate 404
+    @action(methods=["post"], detail=True, url_path="run")
+    def run_benchmark(self, request, *args, **kwargs):
+        benchmarking_session = get_object_or_404(
```
Have you tested this? Since this is a pure django shortcut function I would guess it will raise `Http404` that would return the default error page (so I guess not a json response?): https://docs.djangoproject.com/en/3.0/topics/http/views/#django.http.Http404

I would do it like this:
```
benchmark_id = kwargs["pk"]
try:
    benchmarking_session = BenchmarkingSession.objects.get(id=benchmark_id)
except BenchmarkingSession.DoesNotExist:
    error = f"Benchmark with id {benchmark_id} could not be found"
    return Response({"detail": error}, status=status.HTTP_404_NOT_FOUND
else:
    #  run the task and return response
    pass
```

neopostmodern replied:
from what i understand django REST framework handles this for us. here's [the doc](https://www.django-rest-framework.org/api-guide/exceptions/).
i've just tested it with CURL and it gave me `{"detail":"Not found."}` so i think it's alright? (if we wanted a more verbose/explicit error message as the one you have in your snippet maybe we should focus on overriding the error handling – as described in the doc – instead of having these manual error handling methods all over the place?)

maccinza commented on mmvb_backend/benchmarking_sessions/api/views.py:
```diff
@@ -0,0 +1,124 @@
+        benchmarking_session.task_id = task.id
+        benchmarking_session.save()
+        return Response(
+            f"/benchmarking-sessions/{benchmarking_session.id}/status",
```
I'll check it out locally to see how the response is returned in this case but I'm personally not a fan of  the content not being attached to a name there.
I would return something like:
```
return Response(
    {"status_url": f"/benchmarking-sessions/{benchmarking_session.id}/status"},
    status=status.HTTP_202_ACCEPTED,
)
```

maccinza commented on mmvb_backend/benchmarking_sessions/api/views.py:
```diff
@@ -0,0 +1,124 @@
+    # todo: document response structure (OpenAPI)
+    @action(methods=["get"], detail=True, url_path="status")
+    def benchmark_status(self, request, *args, **kwargs):
+        benchmarking_session = get_object_or_404(
```
Check above if it makes sense.

maccinza commented on mmvb_backend/benchmarking_sessions/api/views.py:
```diff
@@ -0,0 +1,124 @@
+                status=status.HTTP_200_OK,
+            )
+
+        for case_responses in result.info["responses"]:
```
This is just cleaning up the results, right?
Might be worthy to have a separate reusable function to do the cleaning here.

neopostmodern replied:
sure – where would be the django place to put such a function?
(it's not really "cleaning up" – it's reducing the payload for the status update. there is no interest in the actual responses at this stage, only whether the system received a response or if not, what kind of error)

maccinza replied:
I guess this is not related to a "django pattern" or if there is one I'm not aware. I would check:
  - Would this "reducing" function be more closely related to api views? If so, I would create a `utils.py` inside `benchmarking_sessions/api` and have it implemented there.
  - Or does it have a possible broader usage within this application? If so, I would create a `utils.py` inside `benchmarking_sessions`.

Having it more self-contained like this would help improve reusability, code readability, and also make it easier to implement more self-contained unit tests.

maccinza commented on mmvb_backend/benchmarking_sessions/api/views.py:
```diff
@@ -0,0 +1,124 @@
+    # todo: document response structure (OpenAPI)
+    @action(methods=["get"], detail=True, url_path="results")
+    def benchmark_results(self, request, *args, **kwargs):
+        benchmarking_session = get_object_or_404(
```
Same as above if it makes sense.

maccinza commented on mmvb_backend/benchmarking_sessions/api/views.py:
```diff
@@ -0,0 +1,124 @@
+        ).data
+
+        # todo: add metrics infrastructure, add more metrics
+        proportion_cases_with_ai_result_counts = {}
```
Maybe it is just me but this block looks a little bit hard to read. I would suggest:
```
from collections import defaultdict

proportion_cases_with_ai_result_counts = defaultdict(int)
responses = benchmarking_session_result["responses"]

for case_response in responses:
    case_responses = case_response["responses"]
    for ai_implementation_id, response in case_responses.items():
        step_status_value = BenchmarkingStepStatus.COMPLETED.value
        increment = int(response["status"] == step_status_value)
        proportion_cases_with_ai_result_counts[ai_implementation_id] += increment
```

neopostmodern replied:
didn't know about `defaultdict`, neat!
i've given it a try to clean up, please have another look.
(same for below)

maccinza commented on mmvb_backend/benchmarking_sessions/api/views.py:
```diff
@@ -0,0 +1,124 @@
+                    == BenchmarkingStepStatus.COMPLETED.value
+                )
+
+        proportion_cases_with_ai_result_values = {}
```
Would do something similar to what is suggested above to improve readability if it makes sense.

maccinza commented on mmvb_backend/benchmarking_sessions/migrations/0001_initial.py:
```diff
@@ -0,0 +1,77 @@
+                    "status",
+                    models.CharField(
+                        choices=[
+                            ("created", "Created"),
```
`intermediate` was not generated as a choice on the migrations? 🤔 
Maybe it was added later and another migration is missing?

neopostmodern replied:
🤦

Commit:
refactor: address PR feedback

- benchmarking session `/run` now returns the status URL in a JSON object
- cleaned up the preliminary metric code
- include migration for `INTERMEDIATE` benchmarking session status

Commit:
Merge branch 'refactor-to-django' into django-benchmarking-session

maccinza commented on mmvb_backend/benchmarking_sessions/api/views.py:
```diff
@@ -0,0 +1,126 @@
+            )
+
+        result = run_benchmark.AsyncResult(benchmarking_session.task_id)
+        if result.status == "PENDING" or result.info is None:
```
Maybe it would be a good idea using `BenchmarkingStepStatus` instead of the plain string here.

neopostmodern replied:
i'll refer to the celery constant instead, which i think is the most correct

maccinza commented on mmvb_backend/benchmarking_sessions/api/views.py:
```diff
@@ -0,0 +1,126 @@
+
+        benchmarking_session_result["aggregatedMetrics"] = {
+            "proportion_cases_with_ai_result": {
+                "id": "proportion_cases_with_ai_result",
```
Isn't the `id` here kind of redundant?
`aggregatedMetrics` could be either a list with multiple `{"id": "ID", "name": "NAME", "values": []}`
or still be a dictionary with the `id` as key and `{"name": "NAME", "values": []}` as `value`.
What do you think?

neopostmodern replied:
it is often quite nice to have the values as key-object maps in the frontend, but i see how it's not something we want to tackle in the backend. (besides there is a utility function in the frontend already anyway for that purpose).

maccinza commented on mmvb_backend/benchmarking_sessions/migrations/0002_auto_20200514_0933.py:
```diff
@@ -0,0 +1,27 @@
+# Generated by Django 3.0.4 on 2020-05-14 09:33
```
This is just a general tip, don't worry about it for this PR.
When dealing with migrations we usually try to keep their numbers as low as possible, so if a a set of migrations haven't still been merged into the main branch and are still under a dev branch we usually try to [squash](https://docs.djangoproject.com/en/3.0/topics/migrations/#squashing-migrations) them. It can be a tricky process, though.

maccinza commented on mmvb_backend/benchmarking_sessions/tasks.py:
```diff
@@ -0,0 +1,161 @@
+@shared_task(bind=True)
+def run_benchmark(self, benchmarking_session_id):
+    # todo: make configurable in benchmarking session
+    TIMEOUT = 10
```
Maybe instead of configuring it on the benchmarking session it could be configured in the settings (and even loaded from an environment variable).

neopostmodern replied:
i guess we'll need to discuss this. i could imagine that for different test scenarios different values might be necessary. (e.g. development benchmarks with ~10 seconds, "real" benchmarks with only one second to prevent tampering/adjustments)
but the default value should probably be part of the settings. 
@maccinza can you maybe move the value to settings? you're probably much faster at it than i would be (with figuring out how the settings really work)

maccinza replied:
Sure, I can do that one.
In my opinion, in the future, maybe it would be good to have a `BenchmarkingConfiguration` model where those kind of settings could be configured and when creating a `BenchmarkingSession` it should reference a `BenchmarkingConfiguration`.
This would make those configurations more dynamic and accessible to users.

maccinza commented on mmvb_backend/benchmarking_sessions/tasks.py:
```diff
@@ -0,0 +1,161 @@
+            request_ai_implementation_map[request] = ai_implementation
+
+        # wait for all to complete
+        for request in as_completed(request_futures):
```
Nice one!

Commit:
Moves timeout configuration to settings

Commit:
refactor: address PR feedback, part 2

- use celery constant for "PENDING" state
- return `aggregatedMetrics` as array rather than id-object map
- extract value removal for results into a utility function

maccinza reviewed (approved):
Amazing work!
